charms/worker/k8s/src/charm.py

```diff
@@ -227,6 +231,35 @@ def is_worker(self) -> bool:
         """Returns true if the unit is a worker."""
         return self.meta.name == "k8s-worker"
 
+    @property
+    def datastore(self) -> str:
+        """Return the datastore type."""
+        return str(self.config.get("bootstrap-datastore"))
+
+    def get_worker_versions(self) -> Dict[str, List[ops.Unit]]:
+        """Get the versions of the worker units.
+
+        Returns:
+            Dict[str, List[ops.Unit]]: A dictionary of versions and the units that have them.
+        """
+        if not (relation := self.model.get_relation("k8s-cluster")):
+            return {}
+
+        versions = defaultdict(list)
+        for unit in relation.units:
+            if version := relation.data[unit].get("version"):
+                versions[version].append(unit)
+        return versions
+
+    def grant_upgrade(self):
+        """Grant the upgrade to the charm."""
+        self._upgrade_snap = True
+
+    @property
+    def is_upgrade_granted(self) -> bool:
+        """Check if the upgrade has been granted."""
+        return self._upgrade_snap
```

Comment:
Not used

Reply:
Nvm, just saw it's used.

Reply:
Used here: https://github.com/canonical/k8s-operator/blob/7e7e480eb5a5fe53f366f26231fe8995d7f39d0c/charms/worker/k8s/src/snap.py#L276